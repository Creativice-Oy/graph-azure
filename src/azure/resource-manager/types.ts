import { ServiceClientCredentials } from '@azure/ms-rest-js';

/**
 * Credentials obtained for accessing Azure Resource Manager APIs.
 */
export interface AzureManagementClientCredentials {
  credentials: ServiceClientCredentials;
  subscriptionId: string;
}
