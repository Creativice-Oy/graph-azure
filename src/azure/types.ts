import * as MicrosoftGraph from "@microsoft/microsoft-graph-types";

export enum MemberType {
  USER = "#microsoft.graph.user",
  GROUP = "#microsoft.graph.group",
}

export interface Group extends MicrosoftGraph.Group {
  id: string;
  classification: any;
}

export interface User extends MicrosoftGraph.User {
  id: string;
}

export interface GroupMember {
  id: string;
  "@odata.type": string;
  [key: string]: any;
}

export interface GroupMembers {
  group: Group;
  members: GroupMember[];
}

export interface AzureDataModel {
  groups: Group[];
  users: User[];
  groupsMembers: GroupMembers[];
}
