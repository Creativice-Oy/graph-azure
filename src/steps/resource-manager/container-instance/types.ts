import { Container, Volume } from '@azure/arm-containerinstance/esm/models';

export type ContainerWithId = Container & { id: string };
export type VolumeWithId = Volume & { id: string };
