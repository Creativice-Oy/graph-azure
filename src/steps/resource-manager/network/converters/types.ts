export type SecurityRulePortRange = {
  /**
   * The original port range value.
   */
  portRange: string;

  fromPort?: number;
  toPort?: number;
};
