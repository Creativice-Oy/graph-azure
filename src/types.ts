import {
  GraphClient,
  IntegrationExecutionContext,
  PersisterClient,
} from "@jupiterone/jupiter-managed-integration-sdk";
import { AzureClient, ResourceManagerClient } from "./azure";

/**
 * Properties provided by the `IntegrationInstance.config`. Values identifying
 * the Service Principal are included.
 *
 * The Service Principal is used for both Microsoft Graph API and Azure
 * Resource Manager API invocations. The user must grant the Principal access
 * to necessary Graph and Azure resources.
 */
export interface AzureIntegrationInstanceConfig {
  /**
   * The Service Principal client identifier used to obtain API access tokens.
   */
  clientId: string;

  /**
   * The Service Principal client secret used to obtain API access tokens.
   */
  clientSecret: string;

  /**
   * The domain or tenant ID containing the Service Principal used to obtain API
   * access tokens AND to target for ingesting Microsoft Graph resources.
   *
   * The current expectation is that an App Registration, restricted to the
   * tenant it belongs to, is created in the target tenant, and the target
   * tenant has had an Administrator grant read access to Microsoft Graph
   * resources ingested by the program.
   */
  directoryId: string;

  /**
   * The Azure Subscription from which to fetch Azure resources. Currently not
   * guaranteed to be present in all integration instances.
   *
   * A Subscription Administrator must assign the "Reader" RBAC role (least
   * privilege) to the Service Principal used to obtain API access tokens.
   */
  subscriptionId?: string;
}

export interface AzureExecutionContext extends IntegrationExecutionContext {
  graph: GraphClient;
  persister: PersisterClient;
  azure: AzureClient;
  azrm: ResourceManagerClient;
}

export interface ResourceCacheState {
  resourceFetchCompleted?: boolean;
}

export interface GroupsCacheState extends ResourceCacheState {
  groupMembersFetchCompleted?: boolean;
}

export type UsersCacheState = ResourceCacheState;
