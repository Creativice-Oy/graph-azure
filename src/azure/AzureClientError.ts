import { Response } from "node-fetch";

export default class AzureClientError extends Error {
  private static generateMessage(response: Response): string {
    const intro = "Unexpected response from Azure API:";
    const errorText = `'${response.status} - ${response.statusText}`;

    return [intro, errorText].join(" ");
  }

  public status: number;
  public statusText: string;

  constructor(response: Response) {
    super(AzureClientError.generateMessage(response));
    this.status = response.status;
    this.statusText = response.statusText;
  }
}
