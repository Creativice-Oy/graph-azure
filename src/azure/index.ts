import createAzureWebLinker from "./createAzureWebLinker";

export { createAzureWebLinker };

export * from "./types";
