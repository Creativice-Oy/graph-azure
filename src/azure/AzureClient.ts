import querystring from "querystring";

import { IntegrationLogger } from "@jupiterone/jupiter-managed-integration-sdk";

import AzureClientError from "./AzureClientError";
import authenticate from "./graph/authenticate";
import { Group, GroupMember, User } from "./types";

export enum Method {
  GET = "get",
  POST = "post",
}

export interface PaginationOptions {
  limit?: number;
  nextLink?: string;

  /**
   * The property names for `$select` query param.
   */
  select?: string | string[];
}

interface ResourceUrlOptions extends PaginationOptions {
  path: string;
}

interface FetchResourcesInput {
  url: string;
}

export interface FetchResourcesResponse<T extends microsoftgraph.Entity> {
  nextLink: string | undefined;
  resources: T[];
  err?: AzureClientError;
}

export default class AzureClient {
  private clientId: string;
  private clientSecret: string;
  private directoryId: string;
  private accessToken: string;
  private apiHost = "https://graph.microsoft.com";
  private apiVersion = "v1.0";
  private logger: IntegrationLogger;

  constructor(
    clientId: string,
    clientSecret: string,
    directoryId: string,
    logger: IntegrationLogger,
  ) {
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.directoryId = directoryId;
    this.logger = logger;
  }

  public async authenticate() {
    this.accessToken = await authenticate({
      clientId: this.clientId,
      clientSecret: this.clientSecret,
      directoryId: this.directoryId,
    });
  }

  public async fetchUsers(
    options?: PaginationOptions,
  ): Promise<FetchResourcesResponse<User>> {
    return this.fetchResources({
      url: this.resourceUrl({ path: "/users", ...options }),
    });
  }

  public async fetchGroups(
    options?: PaginationOptions,
  ): Promise<FetchResourcesResponse<Group>> {
    return this.fetchResources({
      url: this.resourceUrl({ path: "/groups", ...options }),
    });
  }

  public async fetchGroupMembers(
    groupId: string,
    options?: PaginationOptions,
  ): Promise<FetchResourcesResponse<GroupMember>> {
    return this.fetchResources({
      url: this.resourceUrl({
        path: `/groups/${groupId}/members`,
        ...options,
      }),
    });
  }

  private async fetchResources<T>(
    input: FetchResourcesInput,
  ): Promise<FetchResourcesResponse<T>> {
    try {
      this.logger.trace({ url: input.url }, "Fetching Azure resources");

      const response = await this.makeRequest<any>(input.url, Method.GET);

      return {
        nextLink: response["@odata.nextLink"],
        resources: response.value,
      };
    } catch (err) {
      if (err.status === 404) {
        return { nextLink: undefined, resources: [] };
      } else {
        this.logger.warn(
          { err, url: input.url, method: Method.GET },
          "Azure resource request failed",
        );

        return { nextLink: undefined, resources: [], err };
      }
    }
  }

  /**
   * Pagination: https://docs.microsoft.com/en-us/graph/paging
   * Throttling with retry after: https://docs.microsoft.com/en-us/graph/throttling
   * Batching requests: https://docs.microsoft.com/en-us/graph/json-batching
   */
  private async makeRequest<T>(
    url: string,
    method: Method,
    headers?: {},
  ): Promise<T> {
    if (!this.accessToken) {
      await this.authenticate();
    }

    const options: RequestInit = {
      method,
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
        Authorization: `Bearer ${this.accessToken}`,
        ...headers,
      },
    };

    const response = await fetch(url, options);
    if (response.status >= 300) {
      throw new AzureClientError(response);
    }

    return response.json();
  }

  /**
   * Constructs a resource url to include optional query params.
   *
   * `fetch` will itself perform url escaping, so this function must not do so
   * itself. For example, the `url` library has a nice API for constructing
   * parameters, but the `url.toString()` function answers excaped strings,
   * which `fetch` then attempts to escape, making the params invalid.
   */
  private resourceUrl(options: ResourceUrlOptions): string {
    if (options.nextLink) {
      return options.nextLink;
    } else {
      const queryParams: { [key: string]: any } = {};

      if (options.limit) {
        queryParams.$top = options.limit;
      }

      if (options.select) {
        queryParams.$select = Array.isArray(options.select)
          ? options.select.join(",")
          : options.select;
      }

      let url = `${this.apiHost}/${this.apiVersion}${options.path}`;
      if (Object.keys(queryParams).length > 0) {
        url +=
          "?" +
          querystring.stringify(queryParams, undefined, undefined, {
            encodeURIComponent: (s: string) => s,
          });
      }

      return url;
    }
  }
}
