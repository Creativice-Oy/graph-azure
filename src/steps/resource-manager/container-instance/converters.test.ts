import { createAzureWebLinker } from '../../../azure';
import {
  createContainerGroupEntity,
  createContainerEntity,
  createVolumeEntity,
} from './converters';
import { ContainerGroup } from '@azure/arm-containerinstance/esm/models';
import { ContainerWithId, VolumeWithId } from './types';

const webLinker = createAzureWebLinker('something.onmicrosoft.com');

describe('createContainerGroupEntity', () => {
  test('properties transferred', () => {
    const data: ContainerGroup = {
      id:
        '/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group',
      containers: [
        {
          command: [],
          environmentVariables: [],
          image: 'nginx',
          name: 'nginx',
          ports: [{ port: 80, protocol: 'TCP' }],
          resources: { requests: { cpu: 0.5, memoryInGB: 1.5 } },
        },
        {
          command: [],
          environmentVariables: [],
          image: 'microsoft/aci-helloworld:latest',
          name: 'hello-world',
          ports: [{ port: 443, protocol: 'TCP' }],
          resources: { requests: { cpu: 0.5, memoryInGB: 1.5 } },
        },
      ],
      initContainers: [],
      ipAddress: {
        dnsNameLabel: 'j1dev-container-group-dns',
        fqdn: 'j1dev-container-group-dns.eastus.azurecontainer.io',
        ip: '52.186.103.187',
        ports: [
          { port: 80, protocol: 'TCP' },
          { port: 443, protocol: 'TCP' },
        ],
        type: 'Public',
      },
      location: 'eastus',
      name: 'j1dev-container-group',
      osType: 'Linux',
      provisioningState: 'Succeeded',
      restartPolicy: 'Always',
      sku: 'Standard',
      tags: {},
      type: 'Microsoft.ContainerInstance/containerGroups',
    };

    const containerGroupEntity = createContainerGroupEntity(webLinker, data);

    expect(containerGroupEntity).toMatchSnapshot();
    expect(containerGroupEntity).toMatchGraphObjectSchema({
      _class: ['Group'],
    });
  });
});

describe('createContainerEntity', () => {
  test('properties transferred', () => {
    const data: ContainerWithId = {
      id:
        '/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group/containers/nginx',
      command: [],
      environmentVariables: [],
      image: 'nginx',
      name: 'nginx',
      ports: [{ port: 80, protocol: 'TCP' }],
      resources: { requests: { cpu: 0.5, memoryInGB: 1.5 } },
    };

    const containerEntity = createContainerEntity(data);

    expect(containerEntity).toMatchSnapshot();
    expect(containerEntity).toMatchGraphObjectSchema({
      _class: ['Container'],
    });
  });
});

describe('createVolumeEntity', () => {
  test('properties transferred', () => {
    const data: VolumeWithId = {
      id:
        '/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group/volumes/nginx-mount',
      name: 'nginx-mount',
      azureFile: {
        shareName: 'j1dev',
        readOnly: false,
        storageAccountName: 'keionnedj1dev',
      },
    };

    const volumeEntity = createVolumeEntity(data);

    expect(volumeEntity).toMatchSnapshot();
    expect(volumeEntity).toMatchGraphObjectSchema({
      _class: ['Disk'],
    });
  });
});
