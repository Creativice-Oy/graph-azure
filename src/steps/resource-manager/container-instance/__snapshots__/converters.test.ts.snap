// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContainerEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Container",
  ],
  "_key": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group/containers/nginx",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "command": Array [],
        "environmentVariables": Array [],
        "image": "nginx",
        "name": "nginx",
        "ports": Array [
          Object {
            "port": 80,
            "protocol": "TCP",
          },
        ],
        "resources": Object {
          "requests": Object {
            "cpu": 0.5,
            "memoryInGB": 1.5,
          },
        },
      },
    },
  ],
  "_type": "azure_container",
  "createdOn": undefined,
  "displayName": "nginx",
  "id": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group/containers/nginx",
  "image": "nginx",
  "name": "nginx",
}
`;

exports[`createContainerGroupEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Group",
  ],
  "_key": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "containers": Array [
          Object {
            "command": Array [],
            "environmentVariables": Array [],
            "image": "nginx",
            "name": "nginx",
            "ports": Array [
              Object {
                "port": 80,
                "protocol": "TCP",
              },
            ],
            "resources": Object {
              "requests": Object {
                "cpu": 0.5,
                "memoryInGB": 1.5,
              },
            },
          },
          Object {
            "command": Array [],
            "environmentVariables": Array [],
            "image": "microsoft/aci-helloworld:latest",
            "name": "hello-world",
            "ports": Array [
              Object {
                "port": 443,
                "protocol": "TCP",
              },
            ],
            "resources": Object {
              "requests": Object {
                "cpu": 0.5,
                "memoryInGB": 1.5,
              },
            },
          },
        ],
        "id": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group",
        "initContainers": Array [],
        "ipAddress": Object {
          "dnsNameLabel": "j1dev-container-group-dns",
          "fqdn": "j1dev-container-group-dns.eastus.azurecontainer.io",
          "ip": "52.186.103.187",
          "ports": Array [
            Object {
              "port": 80,
              "protocol": "TCP",
            },
            Object {
              "port": 443,
              "protocol": "TCP",
            },
          ],
          "type": "Public",
        },
        "location": "eastus",
        "name": "j1dev-container-group",
        "osType": "Linux",
        "provisioningState": "Succeeded",
        "restartPolicy": "Always",
        "sku": "Standard",
        "tags": Object {},
        "type": "Microsoft.ContainerInstance/containerGroups",
      },
    },
  ],
  "_type": "azure_container_group",
  "createdOn": undefined,
  "displayName": "j1dev-container-group",
  "id": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group",
  "location": "eastus",
  "name": "j1dev-container-group",
  "osType": "Linux",
  "provisioningState": "Succeeded",
  "restartPolicy": "Always",
  "sku": "Standard",
  "type": "Microsoft.ContainerInstance/containerGroups",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group",
}
`;

exports[`createVolumeEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Disk",
  ],
  "_key": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group/volumes/nginx-mount",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "azureFile": Object {
          "readOnly": false,
          "shareName": "j1dev",
          "storageAccountName": "keionnedj1dev",
        },
        "name": "nginx-mount",
      },
    },
  ],
  "_type": "azure_volume",
  "classification": null,
  "createdOn": undefined,
  "displayName": "nginx-mount",
  "encrypted": null,
  "id": "/subscriptions/40474ebe-55a2-4071-8fa8-b610acdd8e56/resourceGroups/j1dev/providers/Microsoft.ContainerInstance/containerGroups/j1dev-container-group/volumes/nginx-mount",
  "name": "nginx-mount",
}
`;
